### PR TITLE
Add finalizers to the operator

### PR DESCRIPTION
**Pull request by courtneypacheco**
The aim of this PR is to add finalizers and related logic to the operator, as part of the level 4 operator cert.

As of right now, the code is very much a WIP and has not been tested thoroughly. I will update the PR when it is in a completed state.

### CODE REVIEW

File: controllers/nodefeaturediscovery_controller.go [hunk 1]
```diff
@@ -303,3 +351,182 @@ func applyComponents() (reconcile.Result, error) {
...
+	if _, err := r.getDaemonSet(ctx, nfdNamespace, masterName); !k8serrors.IsNotFound(err) {
+		return true
+	}
+
+	/*
+	// Attempt to find the worker config
+	if _, err := r.getWorkerConfig(nfd); !k8serrors.IsNotFound(err) {
+		return true
+	}
+	*/
```
Comment (ArangoGutierrez) [inline controllers/nodefeaturediscovery_controller.go:489]: Let's delete non used, commented code

File: controllers/nodefeaturediscovery_controller.go [hunk 2]
```diff
@@ -303,3 +351,182 @@ func applyComponents() (reconcile.Result, error) {
...
+	if err := r.deleteDaemonSet(ctx, nfdNamespace, masterName); err != nil {
+		return err
+	}
+
+	/*
+	// Attempt to delete worker config
+	if err := r.deleteWorkerConfig(nfd); err != nil {
+		return err
+	}
+	*/
```
Comment (ArangoGutierrez) [inline controllers/nodefeaturediscovery_controller.go:424]: Delete non used, commented code

File: controllers/nodefeaturediscovery_controller.go [hunk 3]
```diff
@@ -169,6 +173,50 @@ func (r *NodeFeatureDiscoveryReconciler) Reconcile(ctx context.Context, req ctrl
...
+			removeFinalizer(instance, finalizer)
+			if err := r.Update(ctx, instance); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// Add the defined finalizer as a finalizer to the instance if it does not exist
```
Comment (ArangoGutierrez) [inline controllers/nodefeaturediscovery_controller.go:208]:
```suggestion
	// add finalizer
```
clean and simple comment

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

**Comment by ArangoGutierrez** [comment]
Is this still a WIP, or is it ready for review @courtneypacheco ? 
also there are some Un-attended comments

**Comment by courtneypacheco** [comment]
Issue #192 needs to be fixed before I continue this PR.

**Comment by ArangoGutierrez** [comment]
If you think this PR is ready, please remove the WIP from the title

**Comment by ArangoGutierrez** [comment]
Are we going to UPSTREAM this before merging it DOWN? @courtneypacheco

**Comment by ArangoGutierrez** [comment]
Please rebase on https://github.com/openshift/cluster-nfd-operator/pull/199 and remove WIP

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]
please run 
```bash
go mod tidy
go mod vendor
```

to refresh the vendor on this PR

File: controllers/nodefeaturediscovery_resources.go [hunk 4]
```diff
@@ -205,3 +208,263 @@ func panicIfError(err error) {
...
+
+	// If 'cd' is empty, then the resource does not exist, so it cannot
+	// be deleted
+	if cd == "" {
+		return nil
+	}
+
+	return r.Delete(context.TODO(), wc)
+}
+*/
```
Comment (ArangoGutierrez) [inline controllers/nodefeaturediscovery_resources.go:332]: ?

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: controllers/nodefeaturediscovery_resources.go (hunk 4 above)
Comment (courtneypacheco) [reply controllers/nodefeaturediscovery_resources.go:332]: Ah, yep. Good catch. Let me delete that.

**Comment by ArangoGutierrez** [comment]
is the bot actually running? 
/joke

**Review by ArangoGutierrez — APPROVED** [review]
/lgtm

**Pull request merged**